adapters/clickhouse.go

```diff
@@ -430,24 +431,69 @@ func (ch *ClickHouse) Insert(table *Table, valuesMap map[string]interface{}) err
 	return wrappedTx.DirectCommit()
 }
 
+func (ch *ClickHouse) BulkUpdate(table *Table, objects []map[string]interface{}, deleteConditions *DeleteConditions) error {
+	wrappedTx, err := ch.OpenTx()
+	if err != nil {
+		return err
+	}
+	if err := ch.deleteInTransaction(wrappedTx, table, deleteConditions); err != nil {
+		wrappedTx.Rollback()
+		return err
+	}
+	if err := ch.insertInTransaction(wrappedTx, table, objects); err != nil {
+		wrappedTx.Rollback()
+		return err
+	}
+	return wrappedTx.DirectCommit()
+}
+
+func (ch *ClickHouse) deleteInTransaction(wrappedTx *Transaction, table *Table, deleteConditions *DeleteConditions) error {
+	deleteCondition, values := ch.toDeleteQuery(deleteConditions)
+	deleteQuery := fmt.Sprintf(deleteQueryChTemplate, ch.database, table.Name, deleteCondition)
+	deleteStmt, err := wrappedTx.tx.PrepareContext(ch.ctx, deleteQuery)
+	if err != nil {
+		wrappedTx.Rollback()
```

Comment:
don't rollback transaction in "*inTransaction" methods. It will happen in an upper level

Reply:
👍